File: CodinGame/Easy/Prefix code/Prefix code.py
def read_prefix_codes(n: int) -> dict:
    prefix_codes = {}
    for _ in range(n):
        code, ascii_value = input().split()
        prefix_codes[code] = chr(int(ascii_value))
    return prefix_codes

def decode_string(encoded_string: str, prefix_codes: dict) -> str:
    decoded_string = ""
    i = 0

    while i < len(encoded_string):
        found = False
        
        for code, char in prefix_codes.items():
            code_length = len(code)
            
            if encoded_string[i:i + code_length] == code:
                decoded_string += char
                i += code_length
                found = True
                break
        
        if not found:
            return f"DECODE FAIL AT INDEX {i}"

    return decoded_string

def main():
    n = int(input())
    prefix_codes = read_prefix_codes(n)
    encoded_string = input().strip()
    result = decode_string(encoded_string, prefix_codes)
    
    print(result)

if __name__ == "__main__":
    main()

File: Easy/Prefix code/Prefix code.cs
using System;
using System.Collections.Generic;

class Program
{
    static Dictionary<string, char> ReadPrefixCodes(int n)
    {
        var prefixCodes = new Dictionary<string, char>();
        for (int i = 0; i < n; i++)
        {
            string[] input = Console.ReadLine().Split();
            string code = input[0];
            char asciiChar = (char)int.Parse(input[1]);
            prefixCodes[code] = asciiChar;
        }
        return prefixCodes;
    }

    static string DecodeString(string encodedString, Dictionary<string, char> prefixCodes)
    {
        string decodedString = "";
        int i = 0;

        while (i < encodedString.Length)
        {
            bool found = false;

            foreach (var kvp in prefixCodes)
            {
                string code = kvp.Key;
                int codeLength = code.Length;

                if (i + codeLength <= encodedString.Length && 
                    encodedString.Substring(i, codeLength) == code)
                {
                    decodedString += kvp.Value;
                    i += codeLength;
                    found = true;
                    break;
                }
            }

            if (!found)
            {
                return $"DECODE FAIL AT INDEX {i}";
            }
        }

        return decodedString;
    }

    static void Main()
    {
        int n = int.Parse(Console.ReadLine());
        var prefixCodes = ReadPrefixCodes(n);
        string encodedString = Console.ReadLine().Trim();
        string result = DecodeString(encodedString, prefixCodes);

        Console.WriteLine(result);
    }
}